app/workers/delete_object_hierarchy_worker.rb

```
@@ -2,148 +2,168 @@
+    reflection = ar_object.class.reflect_on_association(association)
+    case reflection.macro
+    when :has_many
+      # here we keep original hierarchy entry if we still find an associated object
+      dependent = ar_object.public_send(association).public_send(:background_deletion_scope).take
+      dependent ? [hierarchy_association_string, *hierarchy_entries_for(dependent)] : []
+    when :has_one
+      # maximum of one associated so we never keep the original hierarchy entry
+      hierarchy_entries_for ar_object.public_send(association)
+    else
+      raise ArgumentError, "Cannot handle association #{ar_object}:#{association} type #{reflection.macro}"
     end
+  end
 
-    def destroy_later
-      reflection.many? ? destroy_has_many_association : destroy_has_one_association
-    end
+  def compatibility(_object, caller_worker_hierarchy = [], _background_destroy_method = 'destroy')
+    # maybe requeue first object from hierarchy would be adequate and uniqueness should deduplicate jobs
+    hierarchy_root = Array(caller_worker_hierarchy).first
+    return unless hierarchy_root
 
-    attr_reader :main_object, :reflection, :caller_worker_hierarchy
+    object_class, id = hierarchy_root.match(/Hierarchy-([a-zA-Z0-9_]+)-([\d*]+)/).captures
 
-    private
+    raise DoNotRetryError, "background deletion cannot handle #{hierarchy_root}" unless object_class && id
 
-    def destroy_has_many_association
-      main_object.public_send("#{reflection.name.to_s.singularize}_ids").each do |associated_object_id|
-        associated_object = reflection.class_name.constantize.new
-        associated_object.id = associated_object_id
-        delete_associated_object_later(associated_object)
-      end
-    rescue ActiveRecord::UnknownPrimaryKey => exception
-      Rails.logger.info "DeleteObjectHierarchyWorker#perform raised #{exception.class} with message #{exception.message}"
-    end
+    self.class.delete_later object_class.constantize.find(id.to_i)
```

Should we probably throw `DoNotRetryError` if the object is not found? :thinking:

---

yeah, good catch, actually better do not queue if object is not found